File: llm_processor.py
import requests
import json

class LLMProcessor:
    def __init__(self, model_name="llama3", ollama_url="http://localhost:11434/api/generate"):
        self.model_name = model_name
        self.ollama_url = ollama_url
        print(f"LLMProcessor initialized with model '{self.model_name}' and Ollama URL '{self.ollama_url}'")

    def generate(self, prompt: str, system_message: str = "", temperature: float = 0.7, max_tokens: int = 500) -> str:
        headers = {"Content-Type": "application/json"}
        data = {
            "model": self.model_name,
            "prompt": prompt,
            "system": system_message,
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens
            }
        }
        
        try:
            response = requests.post(self.ollama_url, headers=headers, data=json.dumps(data), stream=True)
            response.raise_for_status() # Raise an exception for HTTP errors
            
            full_response_content = ""
            for chunk in response.iter_lines():
                if chunk:
                    decoded_chunk = json.loads(chunk.decode("utf-8"))
                    full_response_content += decoded_chunk.get("response", "")
                    if decoded_chunk.get("done"): # Check if the response is complete
                        break
            return full_response_content.strip()
        except requests.exceptions.ConnectionError:
            print(f"Erro: Não foi possível conectar ao Ollama em {self.ollama_url}. Certifique-se de que o Ollama está rodando e o modelo '{self.model_name}' está baixado.")
            return "[ERRO LLM: Ollama não acessível]"
        except requests.exceptions.RequestException as e:
            print(f"Erro na requisição Ollama: {e}")
            return f"[ERRO LLM: {e}]"
        except json.JSONDecodeError as e:
            print(f"Erro ao decodificar JSON da resposta do Ollama: {e}")
            return f"[ERRO LLM: JSON inválido]"

if __name__ == '__main__':
    # Exemplo de uso (requer Ollama rodando localmente com o modelo 'llama3' baixado)
    # Para baixar o modelo: ollama run llama3
    llm = LLMProcessor(model_name="llama3")
    
    print("\nGerando resposta para um prompt simples:")
    response = llm.generate("Qual é a capital da França?")
    print(f"Resposta do LLM: {response}")

    print("\nGerando resposta com mensagem de sistema:")
    system_msg = "Você é um assistente de IA prestativo e conciso."
    response = llm.generate("Explique a fotossíntese em uma frase.", system_message=system_msg)
    print(f"Resposta do LLM: {response}")

    print("\nGerando resposta com temperatura mais alta (mais criativa):")
    response = llm.generate("Escreva um pequeno poema sobre o mar.", temperature=0.9, max_tokens=100)
    print(f"Resposta do LLM: {response}")

File: llmProcessor.js
const axios = require("axios");

class LLMProcessor {
    constructor(modelName = "llama3", ollamaUrl = "http://localhost:11434/api/generate") {
        this.modelName = modelName;
        this.ollamaUrl = ollamaUrl;
        console.log(`LLMProcessor initialized with model '${this.modelName}' and Ollama URL '${this.ollamaUrl}'`);
    }

    async generate(prompt, systemMessage = "", temperature = 0.7, maxTokens = 500) {
        const headers = { "Content-Type": "application/json" };
        const data = {
            model: this.modelName,
            prompt: prompt,
            system: systemMessage,
            options: {
                temperature: temperature,
                num_predict: maxTokens,
            },
            stream: false, // Request full response, not streaming
        };

        try {
            const response = await axios.post(this.ollamaUrl, data, { headers });
            return response.data.response.trim();
        } catch (error) {
            if (error.code === 'ECONNREFUSED') {
                console.error(`Error: Could not connect to Ollama at ${this.ollamaUrl}. Make sure Ollama is running and the model '${this.modelName}' is downloaded.`);
                return "[LLM ERROR: Ollama not accessible]";
            } else if (error.response && error.response.data) {
                console.error(`Ollama API Error: ${error.response.status} - ${JSON.stringify(error.response.data)}`);
                return `[LLM ERROR: ${error.response.status} - ${error.response.data.error || 'Unknown error'}]`;
            } else {
                console.error(`LLM Request Error: ${error.message}`);
                return `[LLM ERROR: ${error.message}]`;
            }
        }
    }
}

module.exports = LLMProcessor;

// Exemplo de uso (para testes internos do módulo Node.js)
async function testLLMProcessor() {
    console.log("Starting LLMProcessor test...");
    const llm = new LLMProcessor("llama3");

    console.log("\nGenerating response for a simple prompt:");
    let response = await llm.generate("Qual é a capital da França?");
    console.log(`LLM Response: ${response}`);

    console.log("\nGenerating response with system message:");
    const systemMsg = "Você é um assistente de IA prestativo e conciso.";
    response = await llm.generate("Explique a fotossíntese em uma frase.", systemMsg);
    console.log(`LLM Response: ${response}`);

    console.log("\nGenerating response with higher temperature (more creative):");
    response = await llm.generate("Escreva um pequeno poema sobre o mar.", "", 0.9, 100);
    console.log(`LLM Response: ${response}`);
}

// Uncomment to run the test
// testLLMProcessor();
